pkg/controller/iampolicy/iampolicy_controller.go

```diff
@@ -284,13 +327,22 @@ func checkAllClusterLevel(clusterRoleBindingList *v1.ClusterRoleBindingList, clu
 				for _, subject := range clusterRoleBinding.Subjects {
 					if subject.Kind == "User" {
 						usersMap[subject.Name] = true
+					} else if subject.Kind == "Group" {
+						users, err := getGroupMembership(subject.Name)
+						if err != nil {
+							return 0, err
```

Comment:
Is this early return correct? We're in a loop - should we store the error and count the rest of the things (if possible)?

Reply:
I rather return an error now since we cannot accurately determine the number of users with the cluster role and thus the policy should not be evaluated with incomplete data.

Reply:
A partial count could still be used to determine if it's non-compliant. But I agree, we can't use a partial count to decide that it *is* compliant.

Reply:
So what has me hesitant is that the policy violation specifies the number of users which are over the desired amount. It may be odd to have an incorrect number in there. I suppose I can change the policy violation message to say "Number of users with clusteradmin role is at least %s above the specified limit" in the case where it is a violation but an under count may have occurred.

What do you think?

Reply:
Could we end up with a bad group that breaks our policy? bad finalizer maybe?

Reply:
I ended up making it so that the only error returned is if the query fails and it's not because the group doesn't exist:
https://github.com/open-cluster-management/iam-policy-controller/compare/26f775fbb3139724a1c2cebad549c7ef58131817..a95dc35bd2f1045d349757909fcfe3d4ee3ea2c9

I think this should be a pretty safe approach and avoid any "policy breaking".

Reply:
Since it seems like we won't ever be able to exhaustively enumerate the users that have access to the role (because of external IDP groups and things like that), I'm in favor of changing the violation message to say `at least %s ...`.

What I want to avoid is a user being able to create a Group that breaks our counting mechanism, and then that user being able to add users that we *could* count to exceed the limit, all without triggering a noncompliance event. So I think we need a "best effort" count, and a violation message that reflects that.